### PR TITLE
Rule: add noAccessStateInSetState rule and tests

### PR DESCRIPTION
**Pull request by VictorHom**
[rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-access-state-in-setstate.md)

#341

### CODE REVIEW

**Comment by VictorHom** [comment]
might need to add a test as well regarding unary operations

File: packages/@romejs/diagnostics/descriptions/lint.ts [hunk 1]
```diff
@@ -16,6 +16,10 @@ import {buildSuggestionAdvice} from "../helpers";
 import {createDiagnosticsCategory, orJoin} from "./index";
 
 export const lint = createDiagnosticsCategory({
+	REACT_NO_ACCESS_STATE_IN_SET_STATE: {
+		category: "lint/react/noAccessStateInSetState",
+		message: "Avoid using <emphasis>this.state</emphasis> within a <emphasis>this.setState</emphasis> call.",
```
Comment (bitpshr) [inline packages/@romejs/diagnostics/descriptions/lint.ts:21]:
```suggestion
		message: "Avoid using <emphasis>this.state</emphasis> within a <emphasis>this.setState</emphasis> call.",
		advice: [
			{
				type: "log",
				category: "info",
				text: "Batched state calls could result in unexpected errors due to stale state data.",
			},
		],
```

**Comment by bitpshr** [comment]
This looks great. You may want to update this to use the new test harness signature in #531, but I don't think it's necessary before landing this since all rules will eventually be transitioned iteratively.

**Comment by stefanuros** [comment]
Hey, I was looking at this test and I ran into this case which is evaluated as valid atm.
```
this.setState({
	foo: bar,
	value: {
		x: this.state.x
	}
});
```
Should nested `this.state` be caught?

**Comment by ematipico** [comment]
Good catch @stefanuros !

**Pull request merged**